flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java

```diff
@@ -1048,6 +1071,7 @@ protected void handleDataRefreshEvent(DataRefreshEvent<T> event) {
     }
 
     private void handleDetach() {
+        ui = null;
```

Comment:
I think you need to cancel any pending future here too or things will go wrong if/when it returns
```suggestion
        ui = null;
        if (future != null) {
            future.cancel(true);
            future = null;
        }
```

Reply:
Another question is that should a reference to the executor also be nullified - probably not desired as the grid could be just moved.

Anyway this needs to work so that a shared executor service never has tasks (to execute or to finish) that have references to data communicators that are detached already or its state has been invalidated. Thus you need to cancel the future also from `reset()` method and a few other setters.

Reply:
Done